Source/Hypermedia.Client.Extensions/MicrosoftExtensionsCaching/LinkHcoMemoryUserCache.cs

```diff
@@ -0,0 +1,213 @@
+                case CacheScope.AcrossUserContexts:
+                    return this.sharedUserIdentifier;
+            }
+        }
+
+        public void Replace(
+            Uri uri,
+            TLinkHcoCacheEntry oldEntry,
+            TLinkHcoCacheEntry newEntry)
+        {
+            if (oldEntry.CacheScope != newEntry.CacheScope)
+            {
+                var userIdentifier = DetermineUserIdentifier(oldEntry.CacheScope);
+                this.memoryCache.Remove(this.hcoEntryKeyBuilder(userIdentifier, uri));
+            }
+            this.Set(uri, newEntry);
+        }
+
+        private CancellationChangeTokenWrapper GetUserClearToken(object userClearTokenKey)
+        {
+            CancellationChangeTokenWrapper userClearToken;
+            // use the while loop here instead of GetOrAdd because someone else might overwrite the wrapper at the same time, and the local reference might be invalid because of the post eviction dispose.
+            while (!this.memoryCache.TryGetValue(userClearTokenKey, out userClearToken))
+            {
+                CancellationChangeTokenWrapper rootClearToken;
+                while (!this.memoryCache.TryGetValue(this.rootControlTokenKey, out rootClearToken))
+                {
+                    using (var e = this.memoryCache.CreateEntry(this.rootControlTokenKey))
+                    {
+                        e.Value = new CancellationChangeTokenWrapper();
+                        e.PostEvictionCallbacks.Add(
+                            new PostEvictionCallbackRegistration()
+                            {
+                                EvictionCallback = wrapperEvictionDelegate,
+                            });
+                        this.configureRootExpiration?.Invoke(e);
+                    }
+                }
+                using (var e = this.memoryCache.CreateEntry(userClearTokenKey))
+                {
+                    var wrapper = new CancellationChangeTokenWrapper();
+                    e.Value = wrapper;
+                    e.PostEvictionCallbacks.Add(
+                        new PostEvictionCallbackRegistration()
+                        {
+                            EvictionCallback = wrapperEvictionDelegate,
+                        });
+                    e.ExpirationTokens.Add(rootClearToken.Token);
+                    this.configureControlExpiration?.Invoke(e);
+                }
+            }
+
+            return userClearToken;
+        }
+
+        public void Remove(Uri uri)
+        {
+            this.memoryCache.Remove(this.hcoEntryKeyBuilder(this.sharedUserIdentifier, uri));
+            this.memoryCache.Remove(this.hcoEntryKeyBuilder(this.currentUserIdentifier, uri));
+        }
+
+        public void Clear()
+        {
+            this.memoryCache.Remove(this.controlEntryKeyBuilder(this.currentUserIdentifier));
+            this.memoryCache.Remove(this.controlEntryKeyBuilder(this.sharedUserIdentifier));
```

Comment:
Does a user-level clear trigger a shared clear as well? or is that a separate case?
I would say this is okay, because if an action on a cached hco returns Forbidden, then i want to clear that hco from the cache, but I do not with to know or understand if it was in the private or shared cache.